File: COMP-248/COMP248 Assignement 4/src/Classes.java
/***
 * 
 * This file contains the main classes used for "Nancy's 3D Warrior game"
 * 
 */

import java.util.Random;

class Dice  {

    private int die1 = 0;
    private int die2 = 0;

    Dice()  {}

    public int getDie1()    {   return this.die1;}
    public int getDie2()    {   return this.die2;}

    public int rollDice()  {

        Random random = new Random();

        this.die1 = random.nextInt(6) + 1;
        this.die2 = random.nextInt(6) + 1;

        return this.die1 + this.die2;
    }

    public boolean isDouble()   {
        return this.die1 == this.die2;
    }

    public String toString()    {
        return "Die1: " + this.die1 + " Die2: " + this.die2;
    }

}

class Board {

    private int[][][] board;
    public static final int MIN_LEVEL = 3;
    public static final int MIN_SIZE = 3;
    private int level;
    private int size;

    Board() {
        this.level = 3;
        this.size = 4;
        createBoard(this.level, this.size);
    }

    Board(int l, int x) {
        this.level = l;
        this.size = x;
        createBoard(l, x);
    }

    private void createBoard(int _level, int _size)  {
    	
    	this.board = new int[_level][_size][_size];

        for (int i = 0; i < _level; i++)    {
            for (int j = 0; j < _size; j++) {
                for (int k = 0; k < _size; k++)  {

                    int tempSum = i + j + k;

                    if (tempSum % 3 == 0 && tempSum != 0)
                        this.board[i][j][k] = -3;
                    else if (tempSum % 5 == 0 && tempSum != 0)
                        this.board[i][j][k] = -2;
                    else if (tempSum % 7 == 0 && tempSum != 0)
                        this.board[i][j][k] = 2;
                    else
                        this.board[i][j][k] = 0;                    
                }
            }
        }

    }

    public int getLevel()   {   return this.level;}
    public int getSize()    {   return this.size;}

    public int getEnergyAdj(int l, int x, int y)    {
        return this.board[l][x][y];
    }
    
    public String toString()    {

        String master = "";

        for (int i = 0; i < this.board.length; i++)    {
            master += "\nLevel " + i + "\n";
            master += "----------\n";

            for (int j = 0; j < this.board[i].length; j++) {
                for (int k = 0; k < this.board[i][j].length; k++)  {
                    master += this.board[i][j][k] + "\t";
                }
                master += "\n";
            }
        }
        
        master += "\n";

        return master;
    }

}


class Player	{
	
	private String name;
	private int level;
	private int x;
	private int y;
	private int energy;
	
	Player()	{
		this.name = "";
		this.energy = 10;
		this.level = 0;
		this.x = 0;
		this.y = 0;
	}
	
	Player(String _name)	{
		this.name = _name;
		this.energy = 10;
		this.level = 0;
		this.x = 0;
		this.y = 0;
	}
	
	Player(int _l, int _x, int _y)	{
		this.name = "";
		this.level = _l;
		this.x = _x;
		this.y = _y;
	}

	// Copy constructor and method
	Player(Player objToCopy)	{
		
		if (objToCopy == null)	{
			StackTraceElement l = new Exception().getStackTrace()[0];
			System.err.println(String.format("The object to copy is not defined! @ %s.%s:%s", l.getClassName(), l.getMethodName(), l.getLineNumber()));
			
			System.exit(-1);
		}
		
		this.name = objToCopy.getName();
		this.level = objToCopy.getLevel();
		this.x = objToCopy.getX();
		this.y = objToCopy.getY();
		this.energy = objToCopy.getEnergy();
	}
	
	Player copy()	{
		return new Player(this);
	}
	
	// Getters
	public String getName()		{	return this.name;	}
	public int getLevel()		{	return this.level;	}
	public int getX()			{	return this.x;		}
	public int getY()			{	return this.y;		}
	public int getEnergy()		{	return this.energy;	}
	
	
	// Setters
	public void setName(String value)	{	this.name = value;	}
	public void setLevel(int _level)	{	this.level = _level;}
	public void setX(int _x)			{	this.x = _x;		}
	public void setY(int _y)			{	this.y = _y;		}
	public void setEnergy(int value)	{	this.energy = value;}
	
	public void moveTo(Player p)	{
		this.level = p.getLevel();
		this.x = p.getX();
		this.y = p.getY();
	}
	
	public boolean won(Board b)	{
		if (this.level == b.getLevel() && this.x == b.getSize() -1 && this.y == b.getSize() - 1)
			return true;
		else
			return false;
	}
	
	public boolean equals(Player p )	{
		return (this.level == p.getLevel() && this.x == p.getX() && this.y == p.getY());
	}
	
	public String toString()	{
		return String.format("The player \"%s\" is at level %d, x: %d, y: %d and has %d energy", this.name, this.level, this.x, this.y, this.energy);
	}
	
}

File: COMP-248/COMP248 GUI A4/classes.js
/***
 * 
 * 
 */

 const TILE_SIZE = 150;

class Dice  {
    constructor()  {
        this.die1 = 0;
        this.die2 = 0;
    }

    getDie1()    {   return this.die1;}
    getDie2()    {   return this.die2;}

    rollDice()  {

        const random = (min, max) =>    {   return Math.round(Math.random() * (max - min) + min)};

        this.die1 = random(1, 6);
        this.die2 = random(1, 6);

        return this.die1 + this.die2;
	}
	
	sumOfDice()	{
		return this.die1 + this.die2;
	}

    isDouble()   {
        return this.die1 == this.die2;
    }

    toString()    {
        return "Die1: " + this.die1 + " Die2: " + this.die2;
    }

    draw(x, y)  {

        let DICE_W = 100;

        // For Die1
        push();
        translate(-width / 2 + x, -height / 2 + y);
        fill(255);
        box(DICE_W, DICE_W);

        // Text
        fill(0);
        translate(0, 0, DICE_W - 40);
        text(this.die1, 0, 0);

        pop();

        // For Die2
        push();
        let offset = DICE_W + 10;
        translate(-width / 2 + x + offset, -height / 2 + y);
        fill(255);
        box(DICE_W, DICE_W);

        // Text
        fill(0);
        translate(0, 0, DICE_W - 40);
        text(this.die2, 0, 0);

        pop();

    }

}

class Tile  {
    constructor(x, y, w, h)   {
        this.x = x;
        this.y = y;
        this.w = w;
        this.h = h;
        this.d = 30;
        this.energyAdj = "?";
    }

    draw(playersArray, levelID)  {

        // Detect if current level has player to determine the ALPHA
        let levelHasPlayer = false;
        for (let temp of playersArray)  {
            if (levelID == temp.level)  {
                levelHasPlayer = true;
            }
        }        

        push();
        translate(this.x, this.y);

        // Draw tile box
        if (!levelHasPlayer)    {
            fill('rgba(0, 0, 0, 0.1)');
            stroke('rgba(0, 0, 0, 0.1)');
        }
        box(this.w, this.h, this.d);

        // Energy text
        if (!levelHasPlayer)    {
            fill('rgba(0, 0, 0, 0.5)');
            stroke('rgba(0, 0, 0, 0.5)');
        } else  {
            fill(0, 0, 0);
        }        
        translate(0, 0, this.d / 2 + 10);
        text(this.energyAdj, 0, 0);

        // Draw player
        for (let temp of playersArray)  {
            if (this.x == temp.x && this.y == temp.y && levelID == temp.level)  {
                console.log(temp);
                temp.draw(temp.x * TILE_SIZE, temp.y * TILE_SIZE);               
            }
        }

        pop();
    }
}

class Level {
    constructor(size, id)   {
        this.size = size;
        this.data = [];
        this.levelID = id;
        this.tileSize = TILE_SIZE;
    }

    generate()  {

        this.data = [];

        for (let i = 0; i < this.size; i++) {
            this.data[i] = []
            for (let j = 0; j < this.size; j++) {
                let temp = new Tile(i * this.tileSize, j * this.tileSize, this.tileSize, this.tileSize);
                this.data[i][j] = temp;
            }
        }
    }

    draw(playerArray)  {

        if (this.data.length == 0)  {
            this.generate();
        }

        for (let i = 0; i < this.data.length; i++) {
            for (let j = 0; j < this.data[i].length; j++) {
                this.data[i][j].draw(playerArray, this.levelID);
            }
        }
    }
}

class Board {
    constructor(level, size)    {
        this.level = level;
        this.size = size;
        this.data = [];
    }

    generateBoard() {

        this.data = [];

        for (let i = 0; i < this.level; i++)    {
            this.data.push(new Level(this.size, i));
            this.data[i].generate();
            
            // Determine energy
            for (let j = 0; j < this.size; j++)   {
                for (let k = 0; k < this.size; k++)   {
                    let sum = i + j + k;

                    if (sum % 3 == 0 && sum != 0)   {
                        this.data[i].data[j][k].energyAdj = -3;                        
                    } else if (sum % 5 == 0 && sum != 0)    {
                        this.data[i].data[j][k].energyAdj = -2;
                    } else if (sum % 7 == 0 && sum != 0)    {
                        this.data[i].data[j][k].energyAdj = 2;
                    } else  {
                        this.data[i].data[j][k].energyAdj = 0;
                    }

                } 
            }
        }
    }

    draw(playerArray)  {

        if (this.data.length == 0)  {
            this.generateBoard();
        }

        for (let i = 0; i < this.data.length; i++)  {
            push();
            let offset = 200
            translate(i * 100 - offset, i * 100 - offset, i * - offset / 1.5);
            this.data[i].draw(playerArray);
            pop();
         
        }

    }

    getEnergyAdj(l, x, y)   {
        return this.data[l].data[x][y].energyAdj;
    }

    getSize()   {   return this.size;}
    getLevel()  {   return this.level;}
}

class Player    {
    constructor(name, color)   {
        this.name = name;
        this.level = 0;
        this.x = 0;
        this.y = 0;
        this.color = color || {r: 150, g: 50, b: 50};
        this.energy = 10;
    }

    draw(x, y)  {
        push();
        translate(x, y);
        fill(this.color.r, this.color.g, this.color.b);
        rotateX(Math.PI / 2);
        cylinder(30, 50);
        pop();
    }

    copy()  {
        let tempCopy = new Player(this.name);
        tempCopy.x = this.x;
        tempCopy.y = this.y;
        tempCopy.level = this.level;
        tempCopy.energy = this.energy;
        return tempCopy;
    }

    getX()  {   return this.x;  }
    getY()  {   return this.y;  }
    getEnergy() {   return this.energy;}
    getLevel()  {   return this.level;}

    setX(x) {   return this.x = x;}
    setY(y) {   return this.y = y;}
    setEnergy()   {   return this.energy;}
    setLevel()  {return this.level;}

    equals(other)    {
        return (this.level == other.level && this.x == other.x && this.y == other.y);
    }

    moveTo(temp)    {
        this.level = temp.level;
        this.x = temp.x;
        this.y = temp.y;
    }
}